bluetooth: fix bug when destroying tx buffers on disconnected

mjchen0 opened the pull request:
The logic for destroying buffers on the tx_queue of a connection did not make sure that the buffer returned by the tx_data_pull hook was removed from the connection's tx_queue. If the returned buffer len did not "consume" the enter data in the buf, then we must not destroy it because it is still on the tx_queue and will be returned again on the next call to tx_data_pull.

mjchen0 committed:
bluetooth: fix bug when destroying tx queue buffers on disconnect

Channel tx_queue purging on disconnect was inconsistently handled
by the different channels: iso, l2cap, l2cap_br.

iso channels handled purging in the tx_data_pull hook.

l2cap and l2cap_br did the purging in channel delete functions
and did not expect tx_data_pull to be called for a disconnected
channel. Their data_pull functions could return a ptr to a
net_buf that was still on the tx_queue, which is problematic
when the conn tx_processor unrefs the returned buffer resulting
in multiple calls to the buf destroy function.

To make things consistent and correct, remove the code that tries
to purge tx_queues in the tx_processor and only do purging in
the channels themselves when they are deleted/disconnected.

Also refactor and clarify referencing of the net_buf returned
by tx_data_pull. It was confusing who had a reference and
when, which could vary depending on the length of the original
buffer. There are three cases: the buffer length is less
than the tx.mps, greater the mps but less than the mtu so
requiring segementation but not fragmentation, or greater than
both mps and mtu so requiring both segmentation and fragmentation.
The conn layer would increase the refcnt if the length was
greater than the mtu, but not have any awareness of whether
the net_buf was still on the tx_queue or not.

Now it is the tx_data_pull callbacks responsibitity to increment
the reference count if it is returning a pointer to a net_buf
that it is still keeping on the tx_queue for segmentation purposes.
The conn layer will now always transfer that reference into a
fragment view and not conditional it on the length relative to
the mtu, and always decrement the reference to the parent when
the fragment is destroyed.

So there is no risk of decrementing a reference to a net buf
that might still be on a tx_queue, which simplifies error
handling in particular.

Also add error handling paths for when asserts are not enabled.

Signed-off-by: Mike J. Chen <mjchen@google.com>

alwa-nordic reviewed (changes requested):
I don't understand what the bug is. Can you explain more?

Anyway, I'm sure this change is incorrect. The reference that `conn->tx_data_pull` returns must be disposed of or it will cause a memory leak.

mjchen0 commented:
> I don't understand what the bug is. Can you explain more?
> 
> Anyway, I'm sure this change is incorrect. The reference that `conn->tx_data_pull` returns must be disposed of or it will cause a memory leak.

I was seeing some problems when we had a BLE connection get disconnected and started stress testing the connect/disconnect lifecycle.

The problem scenario is as follows:
1. a net_buf that is larger than the MPS is on the tx_queue when the connection is lost.
2. this code tries to flush the tx_queue by calling tx_data_pull() and destroying all the net_bufs returned. it assumes that since it gave an amount of SIZE_MAX, tx_data_pull() will return a net_buf that has been removed from the tx_queue, but this might not be the case. 
3. since the net_buf length is larger than the MPS, it might not be the last segment/PDU. It will always be the last frag of a PDU, but not necessarily the last PDU of the SDU. if it's not the last segment, the buf is still on the tx_queue, and only ptr (with no reference count increment as far as I can tell), is returned.

If destroy_callback() is invoked on the ptr to a net_buf still on the tx_queue, it will leave a freed net_buf still on the tx_queue, which is bad. If somehow that doesn't crash, the next call to tx_data_pull() will return the same ptr again (since that net_buf wasn't removed earlier), and netbuf_unref() will be called again on an already freed net_buf.

alwa-nordic commented:
You are right. There is something strange about `l2cap_data_pull`. It only sometimes returns a reference with ownership, and sometimes it does not move ownership, and nothing about this is documented on `bt_conn.tx_data_pull`.

I'll look more into this. For a proper fix, we will need to establish a rule for `bt_conn.tx_data_pull`. I suspect it should probably always return a reference with ownership.

alwa-nordic reviewed (dismissed):
I have changed my mind. Looks like you got the pattern right after all. There is some weird ownership semantics in the code in `send_buf` and `bt_conn_processor`. I'll accept this patch.

Would you like to add the following documentation or similar to `bt_conn.tx_data_pull` to formalize the API contract?

> Ownership of the return value `ret` varies as follows:
> - If `(ret->len <= *length)` then the return value carries ownership of a reference to the net_buf.
> - Otherwise, the returned reference has lifetime until the next call to this function on this connection.

That's what happens to the best of my understanding after looking at `l2cap_data_pull`, `send_buf` and `bt_conn_processor`. Though I find the line `frag_len = MIN(conn_mtu(conn), len)` especially confusing. Can `conn_mtu(conn) < len`, and won't this break the ownership rules? Hopefully not.

Let's see if other reviewers agree with me.

Hopefully we can refactor this at some point to something simpler.

I also encourage you to write a test to prove the bug and fix and add it out regression tests, but I won't require it.

mjchen0 commented:
I've been digging into this further and thinking about it more. Initially, I was looking for the smallest change to fix the crash I saw on disconnect at the l2cap layer, but based on the feedback so far and my additional inspection of the code, I'm starting to lean towards a bigger but hopefully cleaner change.

Currently, there's a mix of different handling for disconnect by different channel types. The conn tx_processor is clearly written to expect to cleanup tx_queues in the channels by calling the tx_data_pull hook, but the l2cap and l2cap_br implementations don't really expect to be called that way. They instead have tx_queue purging code in bt_l2cap_chan_del()/bt_l2cap_br_chan_del(). l2cap.c also has a special variant called lbt_l2cap_chan_shutdown(), which seems to only be called  for dynamic channels if a credit pdu allocation fails. 

For iso channels, the iso_data_pull() has an explicit check for the disconnected case and purges the iso.txq there, and no other place.

I'm leaning towards a change as follows:
1) change conn tx_processor to only call the tx_data_pull() when it wants to send. It will no longer call it for disconnected case.
2) so l2cap and l2cap_br will rely on their chan_del functions to purge the tx_queue
3) change iso so that the txq is purged bt_iso_chan_disconnected()
4) add comment to the tx_data_pull to clarify it should not be called on disconnected channels

I'll work on updating/creating test cases for iso and l2cap. There doesn't appear to be any BT_CLASSIC tests in bsim. Any pointers for where I can add a test for the l2cap_br case?

mjchen0 committed:
tests: bsim: add early disconnect tests for iso and l2cap

Add early_disconnect test cases for iso/cis and l2cap/stress
to test the bluetooth stack handling of disconnects
while transmit is still in progress.

Signed-off-by: Mike J. Chen <mjchen@google.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -1207,6 +1207,8 @@ static struct net_buf *l2cap_br_ret_fc_data_pull(struct bt_conn *conn, size_t am
 
+	__ASSERT_NO_MSG(conn->state == BT_CONN_CONNECTED);
+
 	/* Leave the PDU buffer in the queue until we have sent all its
```
I think the `__ASSERT` should be added to line 1576 for L2CAP_BR.
https://github.com/zephyrproject-rtos/zephyr/blob/8d2010e1e179a88766dcdc87cfb655588bcb8e24/subsys/bluetooth/host/classic/l2cap_br.c#L1568-L1579

mjchen0 replied:
agreed, my mistake. fixed now.

PavelVPV commented:
> I'm leaning towards a change as follows:
> 
> 1. change conn tx_processor to only call the tx_data_pull() when it wants to send. It will no longer call it for disconnected case.
> 2. so l2cap and l2cap_br will rely on their chan_del functions to purge the tx_queue
> 3. change iso so that the txq is purged bt_iso_chan_disconnected()
> 4. add comment to the tx_data_pull to clarify it should not be called on disconnected channels

I like this.

> I'll work on updating/creating test cases for iso and l2cap. There doesn't appear to be any BT_CLASSIC tests in bsim. Any pointers for where I can add a test for the l2cap_br case?

Perhaps, @lylezhu2012 can help you with this.

jhedberg commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -288,10 +288,20 @@ struct bt_conn {
 	 * Scheduling from which channel to pull (e.g. for L2CAP) is done at the
 	 * upper layer's discretion.
+	 *
+	 * Ownership of the returned net_buf varies as follows:
+	 *   - If the net_buf->len <= *length, then the net_buf has been removed
+	 *     from the tx_queue of the connection and is now owned by the caller.
```
net_buf objects don't really have something you could call ownership. You have *references* to them which can be owned. A reference is created upon initial allocation or explicit `net_buf_ref()` invocations, and destroyed through `net_buf_unref()`. Could you reword the above to be in terms of references instead? That said, couldn't you simplify this by having the callback always return a reference which the caller owns - then it doesn't matter if there's another reference in the tx_queue or not since the caller either way does the same thing (eventually calls `net_buf_unref()`).

jhedberg replied:
Yes, that's what I was trying to underline in my comment, i.e. references have ownership but the object itself does not (if we had locking for net_buf then whoever holds the lock could be considered "the owner" at that moment, but we don't, so..)

mjchen0 replied:
It's become a bigger change than I originally hoped but I think it makes sense to change where the buffer reference is incremented. I now have the tx_data_pull callback increase the reference if it is returning a pointer to a net_buf that it is still keeping on it's tx_queue. So one reference for itself, and one for the caller. The caller (conn.c) now doesn't have to increase the reference count when creating fragments, so a bit simpler there and no different paths depending on whether fragmenting due to mtu limits are hit or not.

I reworded the description of tx_data_pull to be more about the references and not ownership.

mjchen0 commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1052,16 +1048,8 @@ void bt_conn_tx_processor(void)
-		}
-
-		goto exit;
+		/* Purging of conn's tx_queue will happen in bt_l2cap_disconnected */
```
Comment moved.

mjchen0 commented on subsys/bluetooth/host/conn.c:
```diff
@@ -993,10 +993,6 @@ static void destroy_and_callback(struct bt_conn *conn,
 				 bt_conn_tx_cb_t cb,
```
I refactored error handling and when the cb is invoked.

I noticed there were missing error handling paths if asserts were disabled and added those. Now the send_buf() routine handles all error paths in terms of decrementing net_bufs and invoking callback. Conn layer only does the disconnect on error. Otherwise there were again issues with extra calls to net_buf_unref() when certain checks failed with asserts disabled.

mjchen0 commented on subsys/bluetooth/host/iso.c:
```diff
@@ -827,6 +833,14 @@ static struct net_buf *iso_data_pull(struct bt_conn *conn, size_t amount, size_t
+		 * to the net_buf but we're still keeping a reference
+		 * on the txq.
+		 */
+		__maybe_unused struct net_buf *b = net_buf_ref(frag);
```
I moved where the net_buf_ref() is as suggested.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -1630,6 +1634,8 @@ struct net_buf *l2cap_br_data_pull(struct bt_conn *conn, size_t amount, size_t *
 
 		__ASSERT_NO_MSG(found);
 
+		net_buf_unref(pdu);
```
Shouldn't this operate on `q_pdu`? Doing an unref means it's invalid to try to access the pointer later, however the function does just that (even returning its value). The call to `sys_slist_find_and_remove()` should probably operate on the `q_pdu` as well. I realize these pointers have the same values so in practice there's no difference, but in order to avoid bugs I think it's good to keep our reference counting usage semantically correct. Btw, if it didn't cause substantial API breakage, I think APIs that move the reference into the function, like `net_buf_unref()` should take a pointer to a pointer and set the original value to NULL, thereby making it impossible for the calling function to use it anymore.

mjchen0 replied:
done

PavelVPV commented on tests/bsim/bluetooth/host/l2cap/stress/tests_scripts/l2cap_early_disconnect.sh:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+# Copyright (c) 2022 Nordic Semiconductor
```
Is copyright correct?

mjchen0 replied:
fixed

PavelVPV commented on tests/bsim/bluetooth/host/l2cap/stress/tests_scripts/l2cap_early_disconnect.sh:
```diff
@@ -0,0 +1,33 @@
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+# EATT test
```
Could you please add test purpose and procedure? Yes, the neighboring tests don't have test description, but it is not good.

PavelVPV commented on tests/bsim/bluetooth/host/iso/cis/tests_scripts/cis_early_disconnect.sh:
```diff
@@ -0,0 +1,22 @@
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+simulation_id="iso_cis_early_disconnect"
```
Could you please add the test purpose and procedure?
